### ISSUE TITLE
PF-RING Mechanical Turk

### ISSUE THREAD

**Issue by alongir** [kubeshark/kubeshark#1458]
### Context

- Kubeshark can use PF_RING by loading a kernel model file.
- Kernel model must match the kernel minor OS version.
- Our telemetry function records the kernel versions of running Workers.

Manually create kernel models based on a kernel version
For example: `5.10.184-175.731.amzn2.x86_64`

### First drop

**Please complete in a few days**

- [x] Build a kernel module file for `5.10.184-175.731.amzn2.x86_64` and let's test it, to make sure we are all on the same page.

### Second drop 

- [x] Define and document a process to build a kernel module based on a kernel model version. Process should be automated as much as possible.

**While in parallel**:  

- [x] develop the ability to and perform on-demand the creation of kernel model files upon requests (each build is a new task)

### Definition of Done

- [x] An automated process 
- [x] Documentation
- [x] To a point anyone can operate the process based on the documentation

### Follow on tasks

- [x] Research and come up with a better strategy going forward (e.g.  an automatic mechanism using Helm and KMM )

### Researches

https://gist.github.com/tnozicka/ba3c5242ced4ad9a1d4ee69c1fbab24d
https://github.com/kubeshark/worker/blob/master/docker/module-loader/Dockerfile
https://kmm.sigs.k8s.io/documentation/module_loader_image/

**Comment by alongir** [comment]
@corest FYI

**Comment by corest** [comment]
**First drop**


1. Created cluster
2. Deployed test container with build tools
3. Installed kernel-headers for running kernel
4. Built pf-ring module with those headers
5. Copied pf_ring.ko from build container to local fs and created docker with module loader (important - build with stable release)
6. Deployed cert-manager/kmm into cluster
7. Deployed KMM Module CR with container from stage 5, module was loaded from container into node
8. Run kubeshark tap

in workers logs see

```
2023-12-05T22:28:15Z WRN main.go:135 > error="pf_ring: file exists"
2023-12-05T22:28:15Z INF assemblers/tcp_streams_map.go:75 > Using 1000 ms as the close timed out TCP stream channels interval
2023-12-05T22:28:15Z INF source/tcp_packet_source.go:106 > Using PF_RING socket as the capture source
```

**Comment by alongir** [comment]
@corest can you add a description how I can manually create a `pf-ring.ko` file and load it to a specific node using `kubectl`, `ssh` and `cp` commands?

**Comment by MuhaMMadHammD** [comment]
https://github.com/kubeshark/pf-ring-compiler
This link is not working can you provide more info on it

**Comment by corest** [comment]
Here is the doc I've followed to load module in a newly created cluster

# PF_RING

## Overview

PF_RING™ is an advanced Linux kernel module and user-space framework designed for high-speed packet processing. It offers a uniform API for packet processing applications, enabling efficient handling of large volumes of network data.

For comprehensive information on PF_RING™, please visit the [User's Guide]((https://www.ntop.org/guides/pf_ring) and access detailed [API Documentation](http://www.ntop.org/guides/pf_ring_api/files.html).

## Kubeshark support

Kubeshark seamlessly integrates with PF_RING, provided the kernel module is active on the target cluster nodes. In the absence of the PF_RING module, Kubeshark falls back to a default mode (what is this mode?).

## Installing the Kernel Module Management Operator

Deploying the PF_RING kernel module on cluster nodes can be done in several ways. A convenient method for dynamic loading of the PF_RING kernel module is through the [Kernel Module Management (KMM)](https://kmm.sigs.k8s.io) operator.

Follow the [installation guide](https://kmm.sigs.k8s.io/documentation/install/) for detailed instructions on implementing the KMM operator in your cluster.


## Preparing container with the PF_RING kernel module

1. Find the current kernel versions on the nodes

```bash
nodes=$(kubectl get nodes --no-headers -o custom-columns=":metadata.name")
echo $nodes | xargs -I {} kubectl debug node/{} -it --image=busybox -- /bin/sh -c 'uname -r' 2>/dev/null | grep -v "Creating"
```

2. Verify the availability of a containers for your kernel versions:

Kubeshark maintains a repository of containers for popular kernel versions. Below is a list of supported kernels and their corresponding containers:

| Kernel version | Container |
|----------------|-----------|
|5.10.198-187.748.amzn2.x86_64|kubehq/pf-ring-module:5.10.198-187.748.amzn2.x86_64|
|5.10.199-190.747.amzn2.x86_64|kubehq/pf-ring-module:5.10.199-190.747.amzn2.x86_64|
|5.14.0-362.8.1.el9_3.x86_64|kubehq/pf-ring-module:5.14.0-362.8.1.el9_3.x86_64|
|5.15.0-1050-aws|kubehq/pf-ring-module:5.15.0-1050-aws|

If one or more kernel version is not listed, please [submit a request](https://link-to-issue-template) for adding a new version.

## Loading the PF_RING Kernel Module

Load the PF_RING kernel module using the KMM Module custom resource:

```bash
cat <<EOF | kubectl apply -f -
apiVersion: kmm.sigs.x-k8s.io/v1beta1
kind: Module
metadata:
  name: pf-ring
spec:
  moduleLoader:
    container:
      modprobe:
        moduleName: pf_ring
        dirName: /opt
      kernelMappings:
        - regexp: '^.+$'
          containerImage: 'kubehq/pf-ring-module:\${KERNEL_FULL_VERSION}'
  selector:
    kubernetes.io/hostname:
EOF
```

To verify successful module loading on all nodes, inspect the module status:

```bash
# example for a cluster with 2 nodes
kubectl get modules pf-ring -o json | jq .status.moduleLoader
{
  "availableNumber": 2,
  "desiredNumber": 2,
  "nodesMatchingSelectorNumber": 2
}
```

**Comment by corest** [comment]
Manual pf_ring.ko build process: https://github.com/kubeshark/pf-ring-compiler/blob/main/builders/README.md#manual-pf_ringko-build
Container build with new module: https://github.com/kubeshark/pf-ring-compiler/tree/main/modules#build

**Comment by corest** [comment]
PR merged, task completed
